337 multiple values field type

ekraffmiller opened the pull request:
## What this PR does / why we need it:
Hard-coded version of multiple metadata values input, to be modified for the dynamic Create Dataset form

## Which issue(s) this PR closes: 

- Closes #337 

## Special notes for your reviewer:

## Suggestions on how to test this:

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:

## Is there a release notes update needed for this change?:

## Additional documentation:

MellyGray commented on src/sections/create-dataset/useDatasetFormData.tsx:
```diff
@@ -1,22 +1,46 @@
     setFormData(updatedFormData)
     datasetIsValid(updatedFormData)
   }
+  const addField = (path: string, index: number) => {
+    const updatedFormData = _.cloneDeep(formData)
+    const arrayAtPath: DatasetMetadataSubField[] = _.get(
+      updatedFormData,
+      path
+    ) as DatasetMetadataSubField[]
+
+    arrayAtPath.splice(index, 0, { authorName: '' })
+    _.set(updatedFormData, path, arrayAtPath)
+
+    setFormData(updatedFormData)
+    console.log('added field' + JSON.stringify(updatedFormData.metadataBlocks[0].fields['author']))
+  }
+  const removeField = (path: string, index: number) => {
+    const updatedFormData = _.cloneDeep(formData)
+    const arrayAtPath: string[] = _.get(updatedFormData, path) as string[]
+    arrayAtPath.splice(index, 1)
+    _.set(updatedFormData, path, arrayAtPath)
+    setFormData(updatedFormData)
+    // datasetIsValid(updatedFormData)
+  }
```
It appears that this is no longer in use. Could you remove it? I was confused when reviewing the code, as I couldn't understand why this logic appeared in two different places

MellyGray commented on src/sections/create-dataset/AuthorFormGroup.tsx:
```diff
@@ -0,0 +1,71 @@
+  initialAuthorFields: DatasetMetadataSubField[]
+}
+
+export function AuthorFormGroup({ submissionStatus, initialAuthorFields }: AuthorFormGroupProps) {
```
I believe we need to pass updateFormData as a prop and invoke it within handleFieldChange, since this function is responsible for updating the form data

ekraffmiller replied:
thanks for your comments, @MellyGray I made the changes you suggested

MellyGray commented on src/stories/create-dataset/AuthorFormGroup.stories.tsx:
```diff
@@ -0,0 +1,27 @@
+export default meta
+type Story = StoryObj<typeof AuthorFormGroup>
+
+export const Default: Story = {
```
This is how I see the field. I think `Author Name ?` should be in the same line. Also the add button is not aligned with the field.

![image](https://github.com/IQSS/dataverse-frontend/assets/23359572/201a86e2-f79b-4a30-8fd1-8f2d3f80449e)

ekraffmiller replied:
thanks, I fixed the alignment problems

MellyGray reviewed (changes requested):
Looks good 🎉 

I've just added a few suggestions for minor improvements.

Also, should we remove the dynamic field from the design system?

ekraffmiller commented on src/sections/create-dataset/dynamic-fields-buttons/DynamicFieldsButtons.tsx:
```diff
@@ -0,0 +1,38 @@
+}: AddFieldButtonsProps) {
+  return (
+    <div className={styles.container}>
+      <Tooltip placement="top" overlay="Add">
```
I added translation text to createDataset.json, in the future when we have an edit form, maybe that can be refactored into a separate file for both create and edit forms.

ekraffmiller commented on src/sections/create-dataset/AuthorFormGroup.tsx:
```diff
@@ -0,0 +1,78 @@
+  updateFormData: (name: string, value: string | DatasetMetadataSubField[]) => void
+}
+
+export function AuthorFormGroup({
```
done!  I noticed that findByLabelText() wasn't working, I think because the "Author Name" wasn't attached to the input.  So I created another label, 'Name' that is grouped with the input field.  This also matches what we want for compound fields.

ekraffmiller commented:
Thanks @MellyGray I made the changes you suggested. The e2e test is currently failing on get  total dataset counts, which the clear test data PR should fix.

MellyGray reviewed (approved):
LGTM!

GPortas reviewed (commented):
There is a failing e2e test. I'm not sure if this is a known error with an associated issue to fix it or if the error was introduced in this PR.

![e2eeerrr](https://github.com/IQSS/dataverse-frontend/assets/17010447/a4cfeaba-0c08-40a8-8ffa-eb0de55147cd)

I have tested the create dataset form with the new multiple field in storybook and it looks good! But I have found some issues.

https://github.com/IQSS/dataverse-frontend/assets/17010447/44e12853-d685-4ecc-8c07-adca5b747b5d

1) After the initial load, why is it the only field that appears with a red border, if the others are also required?

2) After clicking the "add" button, to add one more author name, the other required fields show their red border.

3) I've been playing with "add" and "delete" buttons. You can see how after removing the second author name when we have three authors, the form is automatically submitted.

ekraffmiller commented:
Thanks @GPortas for finding that form submission bug!  I fixed that, and updated the validation to use the same hook as the other fields, so now I think the behavior is consistent with the rest of the form.
The e2e tests are failing because of race conditions when clearing test data, which I think will be fixed by this PR #360

GPortas reviewed (commented):
Thanks for the addressed changes, @ekraffmiller 

Maybe the following issues are not very relevant now because the field validation is going to change soon. Possibly not to address in this PR, but at least to keep in mind. Please let me know what you think please: 

The second author field does not show a required field message if it is left blank, anyway if we add a third author field and then remove it and leave the second blank, the required field message appears (I reproduce this in the recording below).

https://github.com/IQSS/dataverse-frontend/assets/17010447/4a7aed4d-a90e-4e25-acee-d442e0e9458f

> if you add an author field, but keep it blank, the validation doesn't flag it, but I think that's ok because it is requiring at least one author

In JSF, if you add a multiple field that is required and leave it blank, the required field message is displayed. I think we should replicate this.

![jsfauthors](https://github.com/IQSS/dataverse-frontend/assets/17010447/4c86bc92-cc7a-4b6a-b922-479029be3455)
